187129562/fixbug/application not reviewed

perryzjc opened the pull request:
### [Pivotal Tracker Link][tracker]

<!-- Complete this section filling in the link to a tracker story. -->
[tracker]: https://www.pivotaltracker.com/story/show/187129562

## What this PR does:
<!-- Complete the following sentence: -->

This pull request fixes the status the teacher can see after login to the account for two cases:
1. after being requested more info from an admin, and teacher login back 
-> teacher should see a yellow warning flash message about requested to update info. (change from not_reviewed to info_needed status)
2. after the teacher updates the info (with meaningful changes, i.e. not the same info as last time)
-> teacher should see a red alert flash message about not reviewed (change from info_needed to not_reviewed)

### Include screenshots, videos, etc.
#### Original Problem from previous semester
<img width="807" alt="image" src="https://github.com/cs169/BJC-Teacher-Tracker-App/assets/92573736/b3ea12a5-ede8-4748-9c34-e99a87389b7f">

#### Current effect after fix
##### Case1: Teacher logged in, but has not update info yet
<img width="1329" alt="image" src="https://github.com/cs169/BJC-Teacher-Tracker-App/assets/92573736/02941cb4-c106-4783-af27-01eeef84ee95">

##### Case2: Teacher logged in, with info updated
<img width="807" alt="image" src="https://github.com/cs169/BJC-Teacher-Tracker-App/assets/92573736/b3ea12a5-ede8-4748-9c34-e99a87389b7f">

#### High possibility of original cause of the problem
Teacher db schema has fields other than the info that can be submitted in the form:
<img width="664" alt="image" src="https://github.com/cs169/BJC-Teacher-Tracker-App/assets/92573736/25d8de53-44fe-45ea-be27-2803538300da">

The original code seems to trigger the `before_update` more often  due to the change of other unrelated fields, and thus cause the reset from `info_needed` to `not_reviewed`:
<img width="491" alt="image" src="https://github.com/cs169/BJC-Teacher-Tracker-App/assets/92573736/104db195-f589-46a0-83e6-d17b76317e8a">

#### Who authored this PR?
<!-- Tag the names of any other contributors -->
1. Perry (Jingchao) Zhong
2. Michael Tao


### How should this PR be tested?

* Is there a deploy we can view?
* What do the specs/features test?
1. Test expected behavior when the teacher get a request to update info, and finish the update
    - test the correct status
    - test the alert message and styling
    - test the update without meaningful change
    
* Are there edge cases to watch out for?

#### Are there any complications to deploying this?

<!-- Data migrations, upgrades, etc. -->
No.

### Checklist:

- [ ] Has this been deployed to a staging environment or reviewed by a customer?
- [x] Tag someone for code review (either a coach / team member)
- [x] I have renamed the branch to match PivotTracker's suggested one (necessary for BlueJay) (e.g. `michael/12345-add-new-feature` Any branch name will do as long as the story ID is there. You can use `git checkout -b [new-branch-name]`)

ArushC commented:
@perryzjc @realmichaeltao everything looks good, however, there is this one thing that I think should be changed:

'And I should see a "danger" flash message "Your application is not reviewed. You may update your information. Please check your email for more information." 

I believe that this should not be a red "danger" flash message as this indicates the severity of the application status being not reviewed is much greater than it actually is. Just keep it consistent. Make it be the same yellow-colored "warning" flash message as for the case where the application status is not reviewed. Or if you want the color to be different, at least make it something milder like orange. You might want to check in with @cycomachead regarding this change, as he is the one who originally suggested the color change for the flash.

perryzjc commented:
> @perryzjc @realmichaeltao everything looks good, however, there is this one thing that I think should be changed:
> 
> 'And I should see a "danger" flash message "Your application is not reviewed. You may update your information. Please check your email for more information."
> 
> I believe that this should not be a red "danger" flash message as this indicates the severity of the application status being not reviewed is much greater than it actually is. Just keep it consistent. Make it be the same yellow-colored "warning" flash message as for the case where the application status is not reviewed. Or if you want the color to be different, at least make it something milder like orange. You might want to check in with @cycomachead regarding this change, as he is the one who originally suggested the color change for the flash.

@ArushC  Based on the discussion I had with Professor Ball today, he agrees that `not_reviewed` should be yellow. I just fixed it and changed the relevant cucumber test.

ArushC commented:
LGTM
